### ISSUE TITLE
Infinite loading screen when running on Linux with Proton

### ISSUE THREAD

**Issue by AlpacaRotorvator** [DiplomacyTeam/Bannerlord.Diplomacy#144]
As the title says, whenever I try to launch the game with Diplomacy enabled, it stays in the loading screen indefinitely. Some complementary info that might be of use:

I'm running BL 1.0.2(the issue has been present since at least 1.0.0) together with Glorious Eggroll Proton 7.38, using Mono as my .NET provider rather than Microsoft's. In the past it was possible to run Diplomacy without errors by manually installing MS' .NET 4.7.2 runtimes, but I'm not sure if that's still the case since it's become way more tedious to install it in newer Wine versions.

My load order is:
```
Harmony v2.2.2.140
ButterLib v2.5.1.0
UIExtenderEx v2.3.1.0
MCM v5.3.2.0
[Native Modules]
Diplomacy v1.1.16.0
```

All mods, including Diplomacy itself, were obtained from the workshop.

When starting the game from a command line this is the stack trace that I get:
```
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeInitializationException: The type initializer for 'Diplomacy.PatchTools.PatchManager' threw an exception. ---> System.TypeInitializationException: The type initializer for 'Diplomacy.Patches.FactionManagerPatch' threw an exception. ---> System.ArgumentException: method arguments are incompatible
  at System.Delegate.CreateDelegate (System.Type type, System.Object firstArgument, System.Reflection.MethodInfo method, System.Boolean throwOnBindFailure, System.Boolean allowClosed) [0x002e3] in <e70d6e9587d64cb3abb4b3f99bbf5a0d>:0 
  at System.Delegate.CreateDelegate (System.Type type, System.Reflection.MethodInfo method, System.Boolean throwOnBindFailure) [0x00000] in <e70d6e9587d64cb3abb4b3f99bbf5a0d>:0 
  at System.Delegate.CreateDelegate (System.Type type, System.Reflection.MethodInfo method) [0x00000] in <e70d6e9587d64cb3abb4b3f99bbf5a0d>:0 
  at Diplomacy.Reflect+Method.GetDelegate[TDelegate] () [0x00010] in <779642e68b214b3297f158cb5fd63fab>:0 
  at Diplomacy.Patches.FactionManagerPatch..cctor () [0x0000c] in <779642e68b214b3297f158cb5fd63fab>:0 
   --- End of inner exception stack trace ---
  at (wrapper managed-to-native) System.RuntimeMethodHandle.GetFunctionPointer(intptr)
  at System.RuntimeMethodHandle.GetFunctionPointer () [0x00000] in <e70d6e9587d64cb3abb4b3f99bbf5a0d>:0 
  at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform.GetFunctionPointer (System.Reflection.MethodBase method, System.RuntimeMethodHandle handle) [0x00000] in <255414689d344b1385ae719f6cce8c55>:0 
  at MonoMod.RuntimeDetour.Platforms.DetourRuntimeILPlatform.GetNativeStart (System.Reflection.MethodBase method) [0x00035] in <255414689d344b1385ae719f6cce8c55>:0 
  at MonoMod.RuntimeDetour.DetourHelper.GetNativeStart (System.Reflection.MethodBase method) [0x00005] in <255414689d344b1385ae719f6cce8c55>:0 
  at HarmonyLib.HarmonySharedState.FindReplacement (System.Diagnostics.StackFrame frame) [0x00042] in <255414689d344b1385ae719f6cce8c55>:0 
  at HarmonyLib.Harmony.GetMethodFromStackframe (System.Diagnostics.StackFrame frame) [0x0000e] in <255414689d344b1385ae719f6cce8c55>:0 
  at Bannerlord.ButterLib.ExceptionHandler.CrashReport+<GetAllInvolvedModules>d__22.MoveNext () [0x0011b] in <3a9b9f4f5d8549ed97699d587bbdae50>:0 
  at Bannerlord.ButterLib.ExceptionHandler.CrashReport+<GetAllInvolvedModules>d__22.MoveNext () [0x00098] in <3a9b9f4f5d8549ed97699d587bbdae50>:0 
  at Bannerlord.ButterLib.ExceptionHandler.CrashReport+<GetAllInvolvedModules>d__22.MoveNext () [0x00098] in <3a9b9f4f5d8549ed97699d587bbdae50>:0 
  at System.Collections.Generic.List`1[T].AddEnumerable (System.Collections.Generic.IEnumerable`1[T] enumerable) [0x00059] in <e70d6e9587d64cb3abb4b3f99bbf5a0d>:0 
  at System.Collections.Generic.List`1[T]..ctor (System.Collections.Generic.IEnumerable`1[T] collection) [0x00062] in <e70d6e9587d64cb3abb4b3f99bbf5a0d>:0 
  at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x00018] in <5c5fa49bfcb245569447e7edd431f418>:0 
  at Bannerlord.ButterLib.ExceptionHandler.CrashReport..ctor (System.Exception exception) [0x0005c] in <3a9b9f4f5d8549ed97699d587bbdae50>:0 
  at Bannerlord.ButterLib.ExceptionHandler.ExceptionReporter.Show (System.Exception exception) [0x00024] in <3a9b9f4f5d8549ed97699d587bbdae50>:0 
  at Bannerlord.ButterLib.ExceptionHandler.ExceptionHandlerSubSystem.CurrentDomain_UnhandledException (System.Object sender, System.UnhandledExceptionEventArgs e) [0x0000f] in <3a9b9f4f5d8549ed97699d587bbdae50>:0 
   --- End of inner exception stack trace ---
  at Diplomacy.SubModule.OnSubModuleLoad () [0x00089] in <779642e68b214b3297f158cb5fd63fab>:0 
  at (wrapper dynamic-method) System.Object.lambda_method(System.Runtime.CompilerServices.Closure)
  at Bannerlord.ModuleLoader.SubModuleWrappers.MBSubModuleBaseWrapper.OnSubModuleLoad () [0x0000a] in <569cfea6a5ea499e852fd9939ea26e62>:0 
  at Bannerlord.ModuleLoader.SubModuleWrappers.MBSubModuleBaseListWrapper.OnSubModuleLoad () [0x00014] in <569cfea6a5ea499e852fd9939ea26e62>:0 
  at Bannerlord.ModuleLoader.Bannerlord_Diplomacy.OnSubModuleLoad () [0x0000e] in <569cfea6a5ea499e852fd9939ea26e62>:0 
  at Bannerlord.ModuleLoader.SubModuleWrappers.Patches.MBSubModuleBasePatch.OnSubModuleLoadPostfix (TaleWorlds.MountAndBlade.MBSubModuleBase __instance) [0x0001c] in <569cfea6a5ea499e852fd9939ea26e62>:0 
  at (wrapper dynamic-method) TaleWorlds.MountAndBlade.MBSubModuleBase.TaleWorlds.MountAndBlade.MBSubModuleBase.OnSubModuleLoad_Patch3(TaleWorlds.MountAndBlade.MBSubModuleBase)
  at TaleWorlds.MountAndBlade.Module.InitializeSubModules () [0x0004e] in <0731cdbada3a4107b3ed2a7e8572a108>:0 
  at TaleWorlds.MountAndBlade.Module.LoadSubModules () [0x001ec] in <0731cdbada3a4107b3ed2a7e8572a108>:0 
  at TaleWorlds.MountAndBlade.Module.Initialize () [0x000c9] in <0731cdbada3a4107b3ed2a7e8572a108>:0 
  at ManagedCallbacks.CoreCallbacksGenerated.Module_Initialize (System.Int32 thisPointer) [0x00006] in <b3fa5cecf79c4fa39cab6dde96530e6b>:0 
```

Also, one fact that may or may not be of use is that up until a few months MCM suffered from a similar issue, albeit the error in the stack trace was rather different. Aragas' take on what finally fixed it can be seen [here](https://github.com/Aragas/Bannerlord.MBOptionScreen/issues/267#issuecomment-1297715007)

**Comment by Aragas** [comment]
I see that ButterLib' crash report is being displayed, that must be the reason the game hangs. Can you confirm that there wasn't a second window being created?

**Comment by Aragas** [comment]
Oh, we didn't hit the display stage. Seems that I'll need to fix something on ButterLib's side, but I don't think its the reason it hangs

**Comment by Aragas** [comment]
This concrete exception should be resolved when there will be a new release. Delegate creation is more strict on Mono, but we switched to another implementation of it

**Comment by AlpacaRotorvator** [comment]
Whoah, thanks for the quick response yet again Aragas! Indeed, that particular exception is gone, now I'm getting another one that seems to be pointing the rabbit hole goes a bit deeper:
```
[ERROR] FATAL UNHANDLED EXCEPTION: System.FormatException: Method static System.Boolean TaleWorlds.CampaignSystem.CampaignBehaviors.PlayerTownVisitCampaignBehavior::game_menu_village_hostile_action_on_condition(TaleWorlds.CampaignSystem.GameMenus.MenuCallbackArgs args) cannot be patched. Reason: The type initializer for 'TaleWorlds.CampaignSystem.CampaignBehaviors.PlayerTownVisitCampaignBehavior' threw an exception.
   at MethodInfo HarmonyLib.PatchFunctions.UpdateWrapper(MethodBase original, PatchInfo patchInfo)
   at MethodInfo HarmonyLib.PatchProcessor.Patch()
   at MethodInfo HarmonyLib.Harmony.Patch(MethodBase original, HarmonyMethod prefix, HarmonyMethod postfix, HarmonyMethod transpiler, HarmonyMethod finalizer)
   at void Diplomacy.PatchTools.PatchClass+Patch.Apply(Harmony harmony)
   at new Diplomacy.PatchTools.PatchManager(string harmonyId)
   at void Diplomacy.PatchTools.PatchManager.PatchAll(string harmonyId)
   at void Diplomacy.SubModule.OnSubModuleLoad()
   at void Bannerlord.ModuleLoader.SubModuleWrappers.MBSubModuleBaseWrapper.OnSubModuleLoad()
   at void Bannerlord.ModuleLoader.SubModuleWrappers.MBSubModuleBaseListWrapper.OnSubModuleLoad()
   at void Bannerlord.ModuleLoader.Bannerlord_Diplomacy.OnSubModuleLoad()
   at void Bannerlord.ModuleLoader.SubModuleWrappers.Patches.MBSubModuleBasePatch.OnSubModuleLoadPostfix(MBSubModuleBase __instance)
   at void TaleWorlds.MountAndBlade.Module.InitializeSubModules()
   at void TaleWorlds.MountAndBlade.Module.LoadSubModules()
   at void TaleWorlds.MountAndBlade.Module.Initialize()
```

Seems like this is now getting out of the scope of what can be done inside Diplomacy's code, right? I'm just not sure whether it's something that could be addressed in Harmony, only by Talewords itself or if we're really hitting the limits of what can be done with Mono right now.

**Comment by Aragas** [comment]
I already had a similar issue, should be able to fix that!

**Comment by Aragas** [comment]
> Whoah, thanks for the quick response yet again Aragas! Indeed, that particular exception is gone, now I'm getting another one that seems to be pointing the rabbit hole goes a bit deeper:
> 
> ```
> [ERROR] FATAL UNHANDLED EXCEPTION: System.FormatException: Method static System.Boolean TaleWorlds.CampaignSystem.CampaignBehaviors.PlayerTownVisitCampaignBehavior::game_menu_village_hostile_action_on_condition(TaleWorlds.CampaignSystem.GameMenus.MenuCallbackArgs args) cannot be patched. Reason: The type initializer for 'TaleWorlds.CampaignSystem.CampaignBehaviors.PlayerTownVisitCampaignBehavior' threw an exception.
>    at MethodInfo HarmonyLib.PatchFunctions.UpdateWrapper(MethodBase original, PatchInfo patchInfo)
>    at MethodInfo HarmonyLib.PatchProcessor.Patch()
>    at MethodInfo HarmonyLib.Harmony.Patch(MethodBase original, HarmonyMethod prefix, HarmonyMethod postfix, HarmonyMethod transpiler, HarmonyMethod finalizer)
>    at void Diplomacy.PatchTools.PatchClass+Patch.Apply(Harmony harmony)
>    at new Diplomacy.PatchTools.PatchManager(string harmonyId)
>    at void Diplomacy.PatchTools.PatchManager.PatchAll(string harmonyId)
>    at void Diplomacy.SubModule.OnSubModuleLoad()
>    at void Bannerlord.ModuleLoader.SubModuleWrappers.MBSubModuleBaseWrapper.OnSubModuleLoad()
>    at void Bannerlord.ModuleLoader.SubModuleWrappers.MBSubModuleBaseListWrapper.OnSubModuleLoad()
>    at void Bannerlord.ModuleLoader.Bannerlord_Diplomacy.OnSubModuleLoad()
>    at void Bannerlord.ModuleLoader.SubModuleWrappers.Patches.MBSubModuleBasePatch.OnSubModuleLoadPostfix(MBSubModuleBase __instance)
>    at void TaleWorlds.MountAndBlade.Module.InitializeSubModules()
>    at void TaleWorlds.MountAndBlade.Module.LoadSubModules()
>    at void TaleWorlds.MountAndBlade.Module.Initialize()
> ```
> 
> Seems like this is now getting out of the scope of what can be done inside Diplomacy's code, right? I'm just not sure whether it's something that could be addressed in Harmony, only by Talewords itself or if we're really hitting the limits of what can be done with Mono right now.

Should be fixed with #158
